In [21]:
import random
import time
import heapq
import tracemalloc
import copy
import concurrent.futures
from collections import deque

In [22]:
class EightPuzzle:
    def __init__(self):
        self.state = self._generate_random_state()

    def _generate_random_state(self):
        while True:
            numbers = list(range(9))
            random.shuffle(numbers)
            grid = [numbers[i:i+3] for i in range(0, 9, 3)]
            if self._is_solvable(grid):
                break
        return grid

    def _is_solvable(self, grid):
        flat_list = [num for row in grid for num in row]
        inversions = 0
        for i in range(len(flat_list)):
            for j in range(i + 1, len(flat_list)):
                if flat_list[i] != 0 and flat_list[j] != 0 and flat_list[i] > flat_list[j]:
                    inversions += 1
        return inversions % 2 == 0
    
    def move_to (self, position, direction):
        row, col = position
        new_row, new_col = row, col

        match direction:
            case 'up':
                new_row = row - 1
            case 'down':
                new_row = row + 1
            case 'left':
                new_col = col - 1
            case 'right':
                new_col = col + 1
            case _:
                print("Invalid direction!")
                return
      
        if 0 <= new_row < 3 and 0 <= new_col < 3:
            self.state[row][col], self.state[new_row][new_col] = self.state[new_row][new_col], self.state[row][col]
            return True
        else: 
            return False
        
    def find_zero(self):
        for i in range(3):
            for j in range(3):
                if self.state[i][j] == 0:
                    return (i, j)

🇧🇷 **pt-br:**
A classe `EightPuzzle` representa a estrutura e a lógica do quebra-cabeça 8-puzzle.
Ao instanciar um novo objeto, um estado inicial aleatório é gerado automaticamente por meio do método `_generate_random_state(self)`, que utiliza o método `_is_solvable(self, grid)` para carantir que o estado inicial seja solucionável.
O método `move_to(self, position, direction)` movimenta o espaço vazio (representado pelo número 0) na direção especificada. Ele retorna True se o movimento for válido e realizado com sucesso, ou False caso contrário, mantendo o estado anterior inalterado.
O método `find_zero(self)` localiza e retorna a posição atual do número 0 na matriz que representa o estado do puzzle.

🇺🇸 **en:**
The `EightPuzzle` class represents the structure and logic of the 8-puzzle game.
When a new object is instantiated, an initial random state is automatically generated using the `_generate_random_state(self)` method, which uses the `_is_solvable(self, grid)` method to ensure that the initial state is solvable.
The `move_to(self, position, direction)` method moves the blank space (represented by the number 0) in the specified direction. It returns True if the move is valid and successfully executed, or False otherwise, keeping the previous state unchanged.
The `find_zero(self)` method locates and returns the current position of the number 0 in the matrix that represents the puzzle state.

In [23]:
class EightPuzzle:
    def __init__(self):
        self.state = self._generate_random_state()

    def _generate_random_state(self):
        while True:
            numbers = list(range(9))
            random.shuffle(numbers)
            grid = [numbers[i:i+3] for i in range(0, 9, 3)]
            if self._is_solvable(grid):
                break
        return grid

    def _is_solvable(self, grid):
        flat_list = [num for row in grid for num in row]
        inversions = 0
        for i in range(len(flat_list)):
            for j in range(i + 1, len(flat_list)):
                if flat_list[i] != 0 and flat_list[j] != 0 and flat_list[i] > flat_list[j]:
                    inversions += 1
        return inversions % 2 == 0
    
    def move_to (self, position, direction):
        row, col = position
        new_row, new_col = row, col

        match direction:
            case 'up':
                new_row = row - 1
            case 'down':
                new_row = row + 1
            case 'left':
                new_col = col - 1
            case 'right':
                new_col = col + 1
            case _:
                print("Invalid direction!")
                return
      
        if 0 <= new_row < 3 and 0 <= new_col < 3:
            self.state[row][col], self.state[new_row][new_col] = self.state[new_row][new_col], self.state[row][col]
            return True
        else: 
            return False
        
    def find_zero(self):
        for i in range(3):
            for j in range(3):
                if self.state[i][j] == 0:
                    return (i, j)

🇧🇷 **pt-br:** A classe `BFSPuzzleAgent` representa um agente que resolve o problema do 8-puzzle utilizando o algoritmo de busca em largura (BFS).
Ao instanciar um novo objeto, a classe recebe um objeto do tipo `EightPuzzle` representando o estado inicial do puzzle. O objetivo do agente é alcançar o estado final:

(1, 2, 3)
(4, 5, 6)
(7, 8, 0)

O método `_state_to_tuple(self, state)` converte o estado do puzzle (uma lista de listas) em uma tupla de tuplas, para que possa ser armazenado e verificado em um conjunto de estados visitados.
O método `solve(self)` implementa o algoritmo BFS para resolver o puzzle. Ele inicia a busca a partir do estado inicial, explorando todos os movimentos possíveis até encontrar o estado objetivo. O algoritmo utiliza uma fila (queue) para explorar os estados em ordem de profundidade. Ele retorna um dicionário com informações sobre o caminho até o objetivo, o custo do caminho, o número de nós expandidos, o tamanho da "franja", o uso máximo de memória, entre outras métricas.

🇺🇸 **en:** The `BFSPuzzleAgent` class represents an agent that solves the 8-puzzle problem using the Breadth-First Search (BFS) algorithm.
When a new object is instantiated, the class receives an `EightPuzzle` object representing the initial state of the puzzle. The goal of the agent is to reach the target state:

(1, 2, 3)
(4, 5, 6)
(7, 8, 0)

The `_state_to_tuple(self, state)` method converts the puzzle state (a list of lists) into a tuple of tuples, so it can be stored and checked in a set of visited states.
The `solve(self)` method implements the BFS algorithm to solve the puzzle. It starts the search from the initial state, exploring all possible moves until the goal state is found. The algorithm uses a queue to explore states in breadth-first order. It returns a dictionary with information about the path to the goal, the cost of the path, the number of nodes expanded, the fringe size, the maximum memory usage, and other metrics.

In [24]:
class BFSPuzzleAgent:
    def __init__(self, puzzle):
        self.puzzle = puzzle
        self.goal_state = (
            (1, 2, 3),
            (4, 5, 6),
            (7, 8, 0)
        )
        self.directions = ['up', 'down', 'left', 'right']

    def _state_to_tuple(self, state):
        return tuple(tuple(row)for row in state)

    def solve(self):
        start_time = time.process_time()
        tracemalloc.start()

        visited = set()
        queue = deque()
        queue.append((copy.deepcopy(self.puzzle.state), [], 0))
        visited.add(self._state_to_tuple(self.puzzle.state))

        nodes_expanded = 0
        max_fringe_size = 1
        max_search_depth = 0
        
        while queue:
            max_fringe_size = max(max_fringe_size, len(queue))
            current_state, path, depth = queue.popleft()

            if self._state_to_tuple(current_state) == self.goal_state:
                end_time = time.process_time()
                current, peak = tracemalloc.get_traced_memory()
                tracemalloc.stop()

                return {
                    'path_to_goal': path,
                    'cost_of_path': len(path),
                    'nodes_expanded': nodes_expanded,
                    'fringe_size': len(queue),
                    'max_fringe_size': max_fringe_size,
                    'search_depth': len(path),
                    'max_search_depth': max_search_depth,
                    'running_time': round(end_time - start_time, 8),
                    'max_ram_usage': round(peak / 1024**2, 8)
                }

            nodes_expanded += 1

            for direction in self.directions:
                temp_puzzle = EightPuzzle()
                temp_puzzle.state = copy.deepcopy(current_state)
                zero_pos = temp_puzzle.find_zero()

                if temp_puzzle.move_to(zero_pos, direction):
                    new_state = temp_puzzle.state
                    state_tuple = self._state_to_tuple(new_state)

                    if state_tuple not in visited:
                        visited.add(state_tuple)
                        queue.append((new_state, path + [direction.capitalize()], depth + 1))
                        max_search_depth = max(max_search_depth, depth + 1)

        return None


🇧🇷 **pt-br:** A classe `DFSPuzzleAgent` representa um agente que resolve o problema do 8-puzzle utilizando o algoritmo de busca em profundidade (DFS).
Ao instanciar um novo objeto, a classe recebe um objeto do tipo `EightPuzzle` representando o estado inicial do puzzle. O objetivo do agente é alcançar o estado final:

(1, 2, 3)
(4, 5, 6)
(7, 8, 0)

O método `_state_to_tuple(self, state)` converte o estado (uma lista de listas) em uma tupla de tuplas, para que possa ser armazenado e verificado em um conjunto de estados visitados.
O método `solve(self, depth_limit=50)` implementa o algoritmo DFS para resolver o puzzle. Ele inicia a busca a partir do estado inicial, explorando todos os movimentos possíveis. A busca ocorre de forma recursiva, com limite de profundidade configurável, até encontrar o estado objetivo ou atingir o limite de profundidade. O algoritmo utiliza uma pilha (stack) para explorar os estados em profundidade. Ele retorna um dicionário com informações sobre o caminho até o objetivo, o custo do caminho, o número de nós expandidos, o tamanho da "franja", o uso máximo de memória, entre outras métricas.

🇺🇸 **en:** The `DFSPuzzleAgent` class represents an agent that solves the 8-puzzle problem using the Depth-First Search (DFS) algorithm.
When a new object is instantiated, the class receives an `EightPuzzle` object representing the initial state of the puzzle. The goal of the agent is to reach the target state:

(1, 2, 3)
(4, 5, 6)
(7, 8, 0)

The `_state_to_tuple(self, state)` method converts the state (a list of lists) into a tuple of tuples, so it can be stored and checked in a set of visited states.

The `solve(self, depth_limit=50)` method implements the DFS algorithm to solve the puzzle. It starts the search from the initial state, exploring all possible moves. The search is recursive, with a configurable depth limit, until the goal state is found or the depth limit is reached. The algorithm uses a stack to explore states in depth-first order. It returns a dictionary with information about the path to the goal, the cost of the path, the number of nodes expanded, the fringe size, the maximum memory usage, and other metrics.

In [25]:
class DFSPuzzleAgent:
    def __init__(self, puzzle):
        self.puzzle = puzzle
        self.goal_state = (
            (1, 2, 3),
            (4, 5, 6),
            (7, 8, 0)
        )
        self.directions = ['up', 'down', 'left', 'right']

    def _state_to_tuple(self, state):
        return tuple(tuple(row) for row in state)

    def solve(self, depth_limit=50):
        start_time = time.process_time()
        tracemalloc.start()

        visited = set()
        stack = [(copy.deepcopy(self.puzzle.state), [], 0)]
        visited.add(self._state_to_tuple(self.puzzle.state))

        nodes_expanded = 0
        max_fringe_size = 1
        max_search_depth = 0

        while stack:
            max_fringe_size = max(max_fringe_size, len(stack))
            current_state, path, depth = stack.pop()

            if self._state_to_tuple(current_state) == self.goal_state:
                end_time = time.process_time()
                current, peak = tracemalloc.get_traced_memory()
                tracemalloc.stop()

                return {
                    'path_to_goal': path,
                    'cost_of_path': len(path),
                    'nodes_expanded': nodes_expanded,
                    'fringe_size': len(stack),
                    'max_fringe_size': max_fringe_size,
                    'search_depth': len(path),
                    'max_search_depth': max_search_depth,
                    'running_time': round(end_time - start_time, 8),
                    'max_ram_usage': round(peak / 1024**2, 8)
                }

            if depth >= depth_limit:
                continue

            nodes_expanded += 1

            for direction in reversed(self.directions):
                temp_puzzle = EightPuzzle()
                temp_puzzle.state = copy.deepcopy(current_state)
                zero_pos = temp_puzzle.find_zero()

                if temp_puzzle.move_to(zero_pos, direction):
                    new_state = temp_puzzle.state
                    state_tuple = self._state_to_tuple(new_state)

                    if state_tuple not in visited:
                        visited.add(state_tuple)
                        stack.append((new_state, path + [direction.capitalize()], depth + 1))
                        max_search_depth = max(max_search_depth, depth + 1)

        return None


🇧🇷 **pt-br:** A classe `IDFSPuzzleAgent` é uma extensão da classe `BFSPuzzleAgent` e implementa o algoritmo de busca em profundidade iterativa (IDFS) para resolver o problema do 8-puzzle.
O método `solve(self, max_depth=50)` realiza a busca iterativa, começando com um limite de profundidade de 0 e aumentando progressivamente até o máximo de `max_depth`. Em cada iteração, o agente executa a busca em profundidade (DFS) até atingir o limite de profundidade especificado. O processo é repetido para profundidades maiores, permitindo que o algoritmo explore o estado do puzzle de maneira iterativa.
O agente retorna um dicionário com informações sobre o caminho até o objetivo, o custo do caminho, o número de nós expandidos, o tamanho da "franja", o uso máximo de memória, entre outras métricas.

🇺🇸 **en:** The `IDFSPuzzleAgent` class is an extension of the `BFSPuzzleAgent` class and implements the Iterative Deepening Depth-First Search (IDFS) algorithm to solve the 8-puzzle problem.
The `solve(self, max_depth=50)` method performs the iterative deepening search, starting with a depth limit of 0 and progressively increasing until the maximum max_depth is reached. In each iteration, the agent performs a depth-first search (DFS) up to the specified depth limit. This process is repeated for increasing depths, allowing the algorithm to explore the puzzle state iteratively.
The agent returns a dictionary with information about the path to the goal, the cost of the path, the number of nodes expanded, the fringe size, the maximum memory usage, and other metrics.

In [26]:
class IDFSPuzzleAgent(BFSPuzzleAgent):
    def solve(self, max_depth=50):
        start_time = time.process_time()
        tracemalloc.start()

        nodes_expanded = 0
        max_fringe_size = 0
        max_search_depth = 0

        for depth_limit in range(max_depth + 1):
            visited = set()
            stack = [(copy.deepcopy(self.puzzle.state), [], 0)]
            visited.add(self._state_to_tuple(self.puzzle.state))

            while stack:
                max_fringe_size = max(max_fringe_size, len(stack))
                current_state, path, depth = stack.pop()

                if self._state_to_tuple(current_state) == self.goal_state:
                    end_time = time.process_time()
                    current, peak = tracemalloc.get_traced_memory()
                    tracemalloc.stop()

                    return {
                        'path_to_goal': path,
                        'cost_of_path': len(path),
                        'nodes_expanded': nodes_expanded,
                        'fringe_size': len(stack),
                        'max_fringe_size': max_fringe_size,
                        'search_depth': len(path),
                        'max_search_depth': max_search_depth,
                        'running_time': round(end_time - start_time, 8),
                        'max_ram_usage': round(peak / 1024**2, 8)
                    }

                if depth >= depth_limit:
                    continue

                nodes_expanded += 1

                for direction in reversed(self.directions):
                    temp_puzzle = EightPuzzle()
                    temp_puzzle.state = copy.deepcopy(current_state)
                    zero_pos = temp_puzzle.find_zero()

                    if temp_puzzle.move_to(zero_pos, direction):
                        new_state = temp_puzzle.state
                        state_tuple = self._state_to_tuple(new_state)
                        if state_tuple not in visited:
                            visited.add(state_tuple)
                            stack.append((new_state, path + [direction.capitalize()], depth + 1))
                            max_search_depth = max(max_search_depth, depth + 1)

        return None


🇧🇷 **pt-br:** A classe `AStarPuzzleAgent` implementa o algoritmo de busca A* para resolver o problema do 8-puzzle. O método `solve(self)` utiliza a função de heurística "distância de Manhattan" para calcular o custo estimado de cada estado até o objetivo. A busca A* é uma combinação de busca em largura com a heurística, ou seja, ela considera tanto o custo real do caminho até o momento quanto a estimativa do custo restante para alcançar a solução. O agente explora os estados do puzzle de maneira otimizada, priorizando os que têm menor custo total estimado. O método retorna um dicionário com informações sobre o caminho até o objetivo, o custo do caminho, o número de nós expandidos, o tamanho da "franja", o uso máximo de memória e outras métricas.

🇺🇸 **en:** The `AStarPuzzleAgent` class implements the A* search algorithm to solve the 8-puzzle problem. The `solve(self)` method uses the "Manhattan distance" heuristic to calculate the estimated cost of each state to the goal. A* search is a combination of breadth-first search with a heuristic, meaning it considers both the actual cost of the path so far and the estimated cost to reach the solution. The agent explores the puzzle states in an optimized way, prioritizing those with the lowest total estimated cost. The method returns a dictionary with information about the path to the goal, the cost of the path, the number of nodes expanded, the fringe size, maximum memory usage, and other metrics.


In [27]:
class AStarPuzzleAgent:
    def __init__(self, puzzle):
        self.puzzle = puzzle
        self.goal_state = ((1,2,3),(4,5,6),(7,8,0))
        self.directions = ['up','down','left','right']

    def _to_tuple(self, state):
        return tuple(tuple(row) for row in state)

    def manhattan(self, state_tuple):
        dist = 0
        for i in range(3):
            for j in range(3):
                v = state_tuple[i][j]
                if v != 0:
                    gi = (v-1)//3
                    gj = (v-1)%3
                    dist += abs(i-gi) + abs(j-gj)
        return dist

    def solve(self):
        start = time.time()
        tracemalloc.start()

        init = self._to_tuple(self.puzzle.state)
        open_list = []
        # (f = g+h, g, state, path)
        heapq.heappush(open_list, (self.manhattan(init), 0, init, []))
        visited = {init}

        nodes_expanded = 0
        max_fringe = 1
        max_depth = 0

        while open_list:
            max_fringe = max(max_fringe, len(open_list))
            f, g, state, path = heapq.heappop(open_list)

            if state == self.goal_state:
                end = time.time()
                _, peak = tracemalloc.get_traced_memory()
                tracemalloc.stop()
                return {
                    'path_to_goal': path,
                    'cost_of_path': len(path),
                    'nodes_expanded': nodes_expanded,
                    'fringe_size': len(open_list),
                    'max_fringe_size': max_fringe,
                    'search_depth': len(path),
                    'max_search_depth': max_depth,
                    'running_time': round(end-start,8),
                    'max_ram_usage': round(peak/1024**2,8)
                }

            nodes_expanded += 1

            # encontra zero
            for i in range(3):
                for j in range(3):
                    if state[i][j] == 0:
                        zr, zc = i, j
                        break
                else:
                    continue
                break

            for d in self.directions:
                new = [list(r) for r in state]
                dr, dc = 0,0
                if d=='up':    dr=-1
                elif d=='down':dr=1
                elif d=='left':dc=-1
                elif d=='right':dc=1
                nr, nc = zr+dr, zc+dc
                if 0<=nr<3 and 0<=nc<3:
                    new[zr][zc], new[nr][nc] = new[nr][nc], new[zr][zc]
                    tup = self._to_tuple(new)
                    if tup not in visited:
                        visited.add(tup)
                        new_g = g + 1
                        new_path = path + [d.capitalize()]
                        max_depth = max(max_depth, len(new_path))
                        f_val = new_g + self.manhattan(tup)
                        heapq.heappush(open_list, (f_val, new_g, tup, new_path))

        tracemalloc.stop()
        return None


🇧🇷 **pt-br:** A classe `GreedyPuzzleAgent` implementa o algoritmo de busca gulosa para resolver o problema do 8-puzzle. O método `solve(self)` utiliza a heurística da "distância de Manhattan" para calcular a distância estimada entre o estado atual e o objetivo. A busca gulosa foca em expandir os estados que estão mais próximos do objetivo, com base na heurística, sem considerar o custo do caminho percorrido até o momento. Isso pode resultar em uma solução mais rápida, mas nem sempre ótima. O agente retorna um dicionário com informações sobre o caminho até o objetivo, o custo do caminho, o número de nós expandidos, o tamanho da "franja", o uso máximo de memória e outras métricas.

🇺🇸 **en:** The `GreedyPuzzleAgent` class implements the Greedy search algorithm to solve the 8-puzzle problem. The `solve(self)` method uses the "Manhattan distance" heuristic to calculate the estimated distance between the current state and the goal. Greedy search focuses on expanding the states that are closest to the goal, based on the heuristic, without considering the cost of the path taken so far. This can result in a faster solution, but not always the optimal one. The agent returns a dictionary with information about the path to the goal, the cost of the path, the number of nodes expanded, the fringe size, maximum memory usage, and other metrics.

In [28]:
class GreedyPuzzleAgent:
    def __init__(self, puzzle):
        self.puzzle = puzzle
        self.goal_state = ((1,2,3),(4,5,6),(7,8,0))
        self.directions = ['up','down','left','right']

    def _to_tuple(self, state):
        return tuple(tuple(row) for row in state)

    def manhattan(self, state_tuple):
        # state_tuple é tupla de tuplas
        dist = 0
        for i in range(3):
            for j in range(3):
                v = state_tuple[i][j]
                if v != 0:
                    gi = (v-1)//3
                    gj = (v-1)%3
                    dist += abs(i-gi) + abs(j-gj)
        return dist

    def solve(self):
        start = time.time()
        tracemalloc.start()

        init = self._to_tuple(self.puzzle.state)
        open_list = []
        heapq.heappush(open_list, (self.manhattan(init), init, []))
        visited = {init}
        
        nodes_expanded = 0
        max_fringe = 1
        max_depth = 0

        while open_list:
            max_fringe = max(max_fringe, len(open_list))
            h, state, path = heapq.heappop(open_list)

            if state == self.goal_state:
                end = time.time()
                _, peak = tracemalloc.get_traced_memory()
                tracemalloc.stop()
                return {
                    'path_to_goal': path,
                    'cost_of_path': len(path),
                    'nodes_expanded': nodes_expanded,
                    'fringe_size': len(open_list),
                    'max_fringe_size': max_fringe,
                    'search_depth': len(path),
                    'max_search_depth': max_depth,
                    'running_time': round(end-start,8),
                    'max_ram_usage': round(peak/1024**2,8)
                }

            nodes_expanded += 1

            # encontra zero
            for i in range(3):
                for j in range(3):
                    if state[i][j] == 0:
                        zr, zc = i, j
                        break
                else:
                    continue
                break

            for d in self.directions:
                # gera sucessor
                new = [list(r) for r in state]
                dr, dc = 0,0
                if d=='up':    dr=-1
                elif d=='down':dr=1
                elif d=='left':dc=-1
                elif d=='right':dc=1
                nr, nc = zr+dr, zc+dc
                if 0<=nr<3 and 0<=nc<3:
                    new[zr][zc], new[nr][nc] = new[nr][nc], new[zr][zc]
                    tup = self._to_tuple(new)
                    if tup not in visited:
                        visited.add(tup)
                        new_path = path + [d.capitalize()]
                        max_depth = max(max_depth, len(new_path))
                        heapq.heappush(open_list, (self.manhattan(tup), tup, new_path))

        tracemalloc.stop()
        return None


🇧🇷 **pt-br**:
A função run_parallel() executa os algoritmos BFS, DFS, IDFS, A* e Greedy em paralelo, utilizando um ThreadPoolExecutor. A função recebe um puzzle como entrada, imprime o estado inicial do quebra-cabeça e cria cinco tarefas paralelas para resolver o problema com cada um dos algoritmos. Ao final, a função imprime os resultados de cada algoritmo, incluindo o caminho até o objetivo, custo, número de nós expandidos, profundidade alcançada, tempo de execução e uso de memória.

🇺🇸 **en**:
The run_parallel() function runs the BFS, DFS, IDFS, A*, and Greedy algorithms in parallel using a ThreadPoolExecutor. The function takes a puzzle as input, prints the initial state of the puzzle, and creates five parallel tasks to solve the problem with each of the algorithms. At the end, the function prints the results of each algorithm, including the path to the goal, cost, number of nodes expanded, search depth, running time, and memory usage.



In [29]:
DEPTH_LIMIT = 1250
MAX_DEPTH = 750

def run_agent(agent_class, puzzle, **kwargs):
    agent = agent_class(copy.deepcopy(puzzle))
    return agent.solve(**kwargs)

def run_parallel(puzzle):
    print("Initial state:")
    for row in puzzle.state:
        print(row)

    start_time = time.time()

    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = {
            'BFS': executor.submit(run_agent, BFSPuzzleAgent, puzzle),
            'DFS': executor.submit(run_agent, DFSPuzzleAgent, puzzle, depth_limit=DEPTH_LIMIT),
            'IDFS': executor.submit(run_agent, IDFSPuzzleAgent, puzzle, max_depth=MAX_DEPTH),
            'A*': executor.submit(run_agent, AStarPuzzleAgent, puzzle),
            'Greedy': executor.submit(run_agent, GreedyPuzzleAgent, puzzle)
        }

        results = {}
        for name, future in futures.items():
            results[name] = future.result()

    end_time = time.time()

    for name, result in results.items():
        print(f"\n--- {name} Result ---")
        if result:
            for key, value in result.items():
                print(f"{key}: {value}")
        else:
            print("No solution found.")

    print(f"\nTotal execution time (parallel): {round(end_time - start_time, 4)}s")

In [30]:
puzzle = EightPuzzle()
run_parallel(puzzle)

Initial state:
[2, 7, 5]
[1, 6, 4]
[3, 0, 8]

--- BFS Result ---
path_to_goal: ['Up', 'Right', 'Down', 'Left', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Left', 'Up', 'Right', 'Down', 'Right', 'Up', 'Left', 'Left', 'Down', 'Down', 'Right', 'Up', 'Right', 'Down']
cost_of_path: 25
nodes_expanded: 153731
fringe_size: 14059
max_fringe_size: 25134
search_depth: 25
max_search_depth: 26
running_time: 33.3125
max_ram_usage: 0.0

--- DFS Result ---
path_to_goal: ['Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Left', 'Down', 'Right', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Right', 'Down', 'Down', 'Left', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 'Up', 'Left', 'Down', 'Down', 'Right', 'Up', 

🇧🇷 **pt-br:**

### Conclusão:  
Os resultados obtidos refletem de forma coerente as características esperadas de cada estratégia de busca aplicada ao 8-puzzle:
- **BFS (Busca em Largura):**  
  Garantiu a solução ótima com custo 25, confirmando sua força em encontrar o caminho mais curto. Em contrapartida, apresentou elevado consumo de recursos, expandindo 153 731 nós e mantendo uma fila com até 25 134 estados, o que resultou em tempo de execução de 33,31 s.
- **DFS (Busca em Profundidade):**  
  Retornou um caminho muito extenso (custo 611), evidenciando sua limitação em optimalidade. Apesar de ser simples, expandiu 117 659 nós, o que demonstra que a exploração profunda e sem limite pode tornar o processo ineficiente.
- **IDFS (Busca em Profundidade Iterativa):**  
  Apresentou boa relação entre qualidade e uso de memória: encontrou solução de custo 35, próximo do ótimo, mas com muito menor ocupação de memória (fringe máxima de 32). Expandiu 341 910 nós em várias iterações, resultando em tempo de 47,05 s.
- **A\* (A-estrela):**  
  Também alcançou o caminho ótimo (25) e se destacou pela eficiência, com apenas 2 834 nós expandidos e tempo inferior a 1 s (0,87 s). A combinação de custo acumulado e heurística manteve a fila controlada (max fringe 1 515).
- **Greedy (Busca Gulosa):**  
  Entregou solução rápida (0,32 s) e consumiu poucos nós (410), mas em custo subótimo (53). Ao focar unicamente na heurística, sacrificou a optimalidade em favor da velocidade.
Em síntese, cada algoritmo se comportou conforme o esperado: BFS e A\* garantem optimalidade — o primeiro com alto custo computacional, o segundo de forma mais equilibrada; DFS falha em optimalidade; IDFS oferece compromisso entre custo e memória; e Greedy prioriza rapidez em detrimento da qualidade da solução.

🇺🇸 **en:**

### Conclusion:  
The results consistently reflect the theoretical profiles of each search strategy applied to the 8-puzzle:
- **BFS (Breadth-First Search):**  
  Secured the optimal solution at cost 25, confirming its strength in finding the shortest path. However, it consumed substantial resources, expanding 153,731 nodes and maintaining a fringe of up to 25,134 states, resulting in 33.31 s of execution time.
- **DFS (Depth-First Search):**  
  Produced a very long path (cost 611), highlighting its lack of optimality. Despite its simplicity, it expanded 117,659 nodes, demonstrating that unbounded deep exploration can be inefficient.
- **IDFS (Iterative Deepening DFS):**  
  Struck a good balance between solution quality and memory use: found a near-optimal solution at cost 35 with a maximum fringe of just 32. It expanded 341,910 nodes over multiple iterations, taking 47.05 s.
- **A\* (A-Star):**  
  Also reached the optimal path (25) and excelled in efficiency, expanding only 2,834 nodes in under a second (0.87 s). The combination of actual cost and heuristic kept the fringe well-controlled (max fringe 1,515).
- **Greedy Search:**  
  Delivered a fast result (0.32 s) with few expansions (410), but at a suboptimal cost (53). By focusing solely on the heuristic, it traded solution quality for speed.
In summary, each algorithm behaved as expected: BFS and A\* guarantee optimality—the former at high computational cost and the latter more balanced; DFS fails on optimality; IDFS offers a compromise between cost and memory; and Greedy prioritizes speed over solution quality.
